### PR TITLE
5721: portal export method

### PR DESCRIPTION
**Pull request by annstriganova**

### CODE REVIEW

File: atsd_client/services.py [hunk 1]
```diff
@@ -674,6 +675,44 @@ def send_commands(self, commands, commit=False):
...
+        del query_params['portal_file']
+        if id is None and name is None:
+            raise ValueError("Either id or name parameter must be specified.")
+
+        possible_themes = ["default", "black"]
+        if theme is not None:
+            if theme.lower() not in possible_themes:
+                raise ValueError("Unsupported theme, use on of: {}".format(", ".join(possible_themes)))
+
+        if portal_file is None:
```
Comment (raipc) [inline atsd_client/services.py]: Isn't it better to use filename from Content-Disposition header as default filename?

```
< Content-Disposition: attachment; filename="ATSD.png"
< Content-Type: image/png;charset=utf-8
< Content-Length: 256763
```

File: atsd_client/services.py [hunk 2]
```diff
@@ -674,6 +675,44 @@ def send_commands(self, commands, commit=False):
...
+        """
+        query_params = locals()
+        del query_params['portal_file']
+        if id is None and name is None:
+            raise ValueError("Either id or name parameter must be specified.")
+
+        possible_themes = ["default", "black"]
+        if theme is not None:
+            if theme.lower() not in possible_themes:
+                raise ValueError("Unsupported theme, use on of: {}".format(", ".join(possible_themes)))
```
Comment (raipc) [inline atsd_client/services.py]: type: use one of

File: atsd_client/services.py [hunk 3]
```diff
@@ -674,6 +675,44 @@ def send_commands(self, commands, commit=False):
         return response
 
 
+# ---------------------------------------------------------------------- PORTAL
+class PortalsService(_Service):
+    def get_portal(self, id=None, name=None, portal_file=None, entity=None, width=900, height=600, theme=None,
+                   additional_params=None):
```
Comment (raipc) [inline atsd_client/services.py]: Won't it be easier for the user to use **kwargs instead of dictionary?

File: atsd_client/services.py [hunk 4]
```diff
@@ -674,6 +675,40 @@ def send_commands(self, commands, commit=False):
...
+        del query_params['portal_file']
+        if id is None and name is None:
+            raise ValueError("Either id or name parameter must be specified.")
+
+        possible_themes = ["default", "black"]
+        if theme is not None:
+            if theme.lower() not in possible_themes:
+                raise ValueError("Unsupported theme, use one of: {}".format(", ".join(possible_themes)))
+
+        if kwargs is not None:
```
Comment (raipc) [inline atsd_client/services.py]: Isn't kwargs always not None?

File: atsd_client/services.py [hunk 5]
```diff
@@ -674,6 +675,40 @@ def send_commands(self, commands, commit=False):
...
+        :param width: `int`  Screenshot width, in pixels. Default: 900.
+        :param height: `int` Screenshot height, in pixels. Default: 600.
+        :param theme: str` Portal theme. Possible values: Default, Black. Default value is set in portal
+        configuration.
+        :param kwargs: `str` Additional request parameters are passed to the target portal
+        and are accessible using the ${parameter_name} syntax.
+        :return: PNG file
+        """
+        query_params = locals()
+        del query_params['portal_file']
```
Comment (raipc) [inline atsd_client/services.py]: kwargs should be removed from query_params

File: atsd_client/services.py (hunk 1 above)
Comment (rodionos) [reply atsd_client/services.py]: I propose the following naming convention for the PNG file:

```
{portal-name}[_{entity_name}]_{yyyymmdd}.png
```

Both portal name and entity name normalized to replace whitespace and unsupported characters with underscore.

ATSD_20181012.png

Docker_portal_nurswghbs001_20181012.png

File: README.md [hunk 6]
```diff
@@ -445,6 +446,15 @@ persist: True
 
 Refer to [API Documentation](https://axibase.com/docs/atsd/api/data/messages/query.html) for additional details.
 
+### Querying Portal
+
+To [export](https://axibase.com/docs/atsd/api/meta/misc/portal.html) portal use `get_portal()` method declared in `PortalsService`:
+
+```python
+ps = PortalsService(connection)
+ps.get_portal(id=192, portal_file="192.png", metrics="property_writes_per_second")
```
Comment (rodionos) [inline README.md]: Add example that passes all arguments: entity, width, height, theme. 

What does `metrics=` parameter stand for?

**Comment by rodionos** [comment]
Sounds good

**Pull request merged**